Feature/add rqtl pairscan

zsloan opened the pull request:
#### Description
This PR will add pair-scan (scantwo) functionality to the rqtl_wrapper script and R/qtl endpoints

#### How should this be tested?
Use the new pair-scan option or just run the rqtl_wrapper.R script with the --pairscan keyword

#### Any background context you want to provide?
<!-- Anything the reviewer should be aware of ahead of testing -->

#### What are the relevant pivotal tracker stories?
<!-- Does this PR track anything anywhere? -->

#### Screenshots (if appropriate)

#### Questions
<!-- Are there any questions for the reviewer -->

zsloan committed:
Added option for running pairscan to rqtl_wrapper.R

zsloan committed:
- Added scan_func function that determines whether scanone or scantwo
(pairscan) is used
- For pairscan default to using step 20 (subject to change, but some
step is required during calc.genoprob to make it run fast enough)
- Added some new verbose prints

zsloan committed:
Added line priting pair-scan results to CSV and changed the default step-size to 10cM for pair-scan

zsloan committed:
Added pairscan boolean kwarg and process_rqtl_pairscan function for reading in pairscan results + renamed process_rqtl_output to process_rqtl_mapping to distinguish between that and pairscan

zsloan committed:
Fix imports to import both process_rqtl_mapping and process_rqtl_pairscan in api/rqtl.py

robwwilliams commented:
Great work Zach.

On Fri, Jul 23, 2021 at 10:15 PM zsloan ***@***.***> wrote:

> Description
>
> This PR will add pair-scan (scantwo) functionality to the rqtl_wrapper
> script and R/qtl endpoints
> How should this be tested?
>
> Use the new pair-scan option or just run the rqtl_wrapper.R script with
> the --pairscan keyword
> Any background context you want to provide? What are the relevant pivotal
> tracker stories? Screenshots (if appropriate) Questions
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/genenetwork/genenetwork3/pull/29
> Commit Summary
>
>    - Added option for running pairscan to rqtl_wrapper.R
>    - - Added scan_func function that determines whether scanone or scantwo
>    - Added line priting pair-scan results to CSV and changed the default
>    step-size to 10cM for pair-scan
>
> File Changes
>
>    - *M* scripts/rqtl_wrapper.R
>    <https://github.com/genenetwork/genenetwork3/pull/29/files#diff-b3e315819e98ca9682934211b2412ffb41d8c1c932c98319daba14b771fb47e4>
>    (41)
>
> Patch Links:
>
>    - https://github.com/genenetwork/genenetwork3/pull/29.patch
>    - https://github.com/genenetwork/genenetwork3/pull/29.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/genenetwork/genenetwork3/pull/29>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC55V7D5K3C4F2LQK7QKXUDTZHEVZANCNFSM5A4UCOBQ>
> .
>
-- 
Rob

Robert W. Williams, Ph.D.
Chair: Department of Genetics, Genomics and Informatics
University of Tennessee Health Science Center
71 S Manassas St, Memphis TN 38103-3310,  TSRB Room 405
Tele/Cell: 901 604 4752
EMAIL: ***@***.***  or  ***@***.***

zsloan committed:
Updated rqtl_wrapper to also return a map file when doing a pair-scan (since we need the list of markers/pseudomarkers and their positions)

zsloan committed:
Create pairscan_for_figure and pairscan_for_table functions that return the Dict and List respectively used for the pair scan figure and the table showing the results

zsloan committed:
Fixed a cople function calls to use the updated function names

zsloan committed:
Removed quotes from beginning and end of chromosome string

zsloan committed:
Added genofile name to inputs for processing R/qtl pair-scan results, since it's needed to store the proximal/distal markers for each position

zsloan committed:
Fix issue that causes R/qtl to always run pair-scan even if pair-scan isn't selected

zsloan committed:
Add functions for getting proximal/distal markers for each pseudomarker position in pair-scan results + return only the sorted top 500 results

zsloan committed:
Add typing to some functions

zsloan committed:
Tried to make the docstrings more consistent

zsloan committed:
Fixed mypy typing errors

zsloan committed:
Fixes pylint errors

Alexanderlacuna committed:
init script ctl analysis

Alexanderlacuna committed:
fetch genotypes and phenotype files

Alexanderlacuna committed:
perform the ctl scan

Alexanderlacuna committed:
get significant interactions

Alexanderlacuna committed:
export json data

Alexanderlacuna committed:
minor fix:read input data from json  file

Alexanderlacuna committed:
generate lineplot

Alexanderlacuna committed:
init file to call ctl script

Alexanderlacuna committed:
add endpoint for ctl

Alexanderlacuna committed:
bugfixes for script;create pheno and geno objects

BonfaceKilz commented:
@zsloan is this good for merging?  If so, could you fix the current merge conflicts so that I can go ahead and merge.

arunisaac committed:
Implement dataset metadata API endpoint.

* guix.scm: Import (gnu packages rdf).
(genenetwork3)[propagated-inputs]: Add python-sparqlwrapper.
* gn3/settings.py (SPARQL_ENDPOINT): New variable.
* gn3/api/general.py: Import datasets from gn3.db.
(dataset_metadata): New API endpoint.
* gn3/db/datasets.py: Import re, Template from string, Dict and Optional from
typing, JSON and SPARQLWrapper from SPARQLWrapper, SPARQL_ENDPOINT from
gn3.settings.
(sparql_query, dataset_metadata): New functions.

arunisaac committed:
Do not recommend GUIX_PACKAGE_PATH.

* guix.scm: Do not recommend GUIX_PACKAGE_PATH.

arunisaac committed:
Replace guix environment with guix shell.

* README.md, guix.scm: Replace guix environment with guix shell.

fredmanglis committed:
Merge pull request #56 from genenetwork/partial-correlations

Partial correlations

fredmanglis committed:
Make the DB connection argument the first

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* To make the code more composable down the line, make the database connection
  argument the first argument for functions that access the database, since
  they will always require the connection.

fredmanglis committed:
Migrate `web.webqtl.correlation.CorrelationPage.fetchAllDatabaseData` function

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Migrate the
  `web.webqtl.correlation.CorrelationPage.fetchAllDatabaseData` function from
  GN1 to GN3.

fredmanglis committed:
Add test to query builders

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Add some tests for the query builders to ensure that the queries are built
  up correctly.

fredmanglis committed:
Update documentation for functions

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Document functions for posterity.

fredmanglis committed:
Migrate `getPartialCorrelationsNormal`

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Migrate the
  `web.webqtl.correlation.PartialCorrDBPage.getPartialCorrelationsNormal`
  function in GN1.
* Remove function obsoleted by newer implementation of the code

fredmanglis committed:
Fix a myriad of linting errors

* Fix linting errors like:

  - Unused variables
  - Undeclared variable errors (mostly caused by typos, and wrong names)
  - Missing documentation strings for functions

  etc.

BonfaceKilz committed:
db: traits: Remove "\n\n" when generating csv file

In excel, "\n\n" is replaced with ",,,," during upload.

BonfaceKilz committed:
db: traits: Remove "\n\n" when generating csv file

In excel, "\n\n" is replaced with ",,,," during upload.

fredmanglis committed:
Retrieve the species name given the group

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Migrate the `web.webqtl.dbFunction.webqtlDatabaseFunction.retrieveSpecies`
  in GeneNetwork1 to `gn3.db.species.species_name` in GeneNetwork3 to enable
  the retrieval of the species name, given the group name (formerly RISet).

BonfaceKilz committed:
db: traits: Remove trailing ".0" in int values

BonfaceKilz committed:
db: traits: Remove trailing ".0" in int values

BonfaceKilz committed:
Add SQL statement to update json_diff

BonfaceKilz committed:
Add SQL statement to update json_diff

BonfaceKilz committed:
db: traits: Support additions and deletions from csv file

BonfaceKilz committed:
db: traits: Support additions and deletions from csv file

fredmanglis committed:
Fix query parametrisation

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Pass parameters to the query the way the MySQL driver expects.

fredmanglis committed:
Update typing notations on functions

Issue: https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

fredmanglis committed:
Return only values

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Return the values from the database, not the tuples.

fredmanglis committed:
Update return type

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Return the complete filename when found, or the boolean value False, when it
  is not found.

fredmanglis committed:
Fix bugs in data

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Process the db_name and db_type values.
* Return data correctly

Alexanderlacuna committed:
fix bug:unpacking error when generator returns empty list

Alexanderlacuna committed:
refactor unittest for normalizing sample values

Alexanderlacuna committed:
refactor sample r unittests

Alexanderlacuna committed:
update function docs for normalizing strain values

Alexanderlacuna committed:
try and catch for non matching sample keys

Alexanderlacuna committed:
rename variables

fredmanglis committed:
Provide entry-point function for the partial correlations

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Provide the entry-point function to the partial correlation feature. This is
  the function that ochestrates the fetching of the data, and processing it
  for output by the API endpoint (to be implemented).

fredmanglis committed:
Fix linting errors

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

fredmanglis committed:
Update test for data changes

Issue: https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

fredmanglis committed:
Merge branch 'main' into partial-correlations

Alexanderlacuna committed:
minor pr fixes

Alexanderlacuna committed:
format normalize  function doc string

arunisaac committed:
Add a Guix operating-system configuration.

* guix-system.scm: New file.

fredmanglis committed:
Merge branch 'partial-correlations'

fredmanglis committed:
Merge branch 'main' of github.com:genenetwork/genenetwork3

fredmanglis committed:
Provide group from primary trait

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* From the collections page, the group is not present, so this commit
  retrieves the group value from the primary trait.

fredmanglis committed:
Prevent error on no result. Fix indexing

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* If the dataset name is not found, don't cause an exception, instead, return
  the provided search name.
* Use the correct inner object

fredmanglis committed:
Provide clearer error message

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* To help with debugging error conditions, provide the original name to help
  with figuring out why the error was raised

arunisaac committed:
Remove pull request template.

People are not using the pull request template properly. Often, the fields
remain unfilled, and the placeholders are not deleted. This is cognitively
taxing for reviewers. Also, most of our changes are better described in the
commit log than in the pull request description. And, our reviews are quite
cursory and don't need a detailed pull request description for the reviewer to
act on.

* .github/PULL_REQUEST_TEMPLATE.md: Delete file.

arunisaac committed:
guix.scm: Inherit from (gn packages genenetwork).

We base this off the genenetwork3 package defined in (gn packages
genenetwork). That way, we don't have to duplicate the package definition in
multiple places.

* guix.scm (genenetwork3): Inherit from genenetwork3 package definition in (gn
packages genenetwork).

arunisaac committed:
README: Add CI badge.

* README.md: Add CI badge.

arunisaac committed:
README: Link CI badge to CI status page.

* README.md: Link CI badge to CI status page.

fredmanglis committed:
Fix parameters

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Provide parameters as a dict of items (which works for MySQL), rather than
  as keywork arguments (as works in PostgreSQL).

fredmanglis committed:
Fix errors with query

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Use the correct format string paradigm: put f before opening quote.
* Provide missing space just before 'FROM' clause

fredmanglis committed:
Run partial correlations against chosen database

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Run the partial correlations against the database that the user selects, and
  not the one from which the primary trait is. This was a bug in the code.

fredmanglis committed:
Return the primary and control traits in addition to results

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* In addition to the partial correlation results, this commit enables the
  return of the chosen primary trait and the selected control traits. This
  data is required for presentation on the results page.

fredmanglis committed:
Fix the removal of controls for corresponding Nones in targets

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Fix the code, so that it removes all control values, whose corresponding
  target values are None, without throwing an error.

fredmanglis committed:
Provide missing function

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Import the missing function.

fredmanglis committed:
Set explicit formatting for sequences in queries

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* the mysqlclient is sometimes finicky with the format specifiers, especially
  in the `WHERE ... IN ...` clauses where a sequence of values is
  provided. This commit explicitly sets the format specifiers for such clauses
  so as to avoid sporadic failures due to differences in the form of data from
  one query to the next.

fredmanglis committed:
Fix dataset: use target dataset not primary

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Use the target dataset to load the target traits, not the primary trait's
  dataset, since they might differ.

fredmanglis committed:
Remove any items with less than 3 samples

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* pingouin raises an exception whenever one attempts to use it to compute the
  partial correlation with data that has less than 3 samples.

fredmanglis committed:
TO REVERT: Add logging to see data frame

fredmanglis committed:
Provide expected values to use for tests

jgarte committed:
Adds type hint for normalize_values function

fredmanglis committed:
mypy: ignore some imports and errors

* Ignore some missing library stubs
* Ignore some typing errors
* Fix obvious typing errors

fredmanglis committed:
Fix bug: expects int, gets object - Fixed.

fredmanglis committed:
linting: Fix obvious linting issues

fredmanglis committed:
Add API endpoint for partial correlations

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Add an API endpoint for the partial correlation.

fredmanglis committed:
Add "success" status to final computation results

Issue: https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

fredmanglis committed:
Add dataset type to the results

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* The dataset type is relevant for the display of the data, therefore, this
  commit presents the dataset type as part of the results.

fredmanglis committed:
Reduce the total amount of data to be output

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* There is a lot of data that is not necessary in the final result. This
  commit removes that data, retaining only data relevant for the display.

fredmanglis committed:
Replace `NaN` with `null` in JSON string

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* `NaN` is not a valid JSON value, and leads to errors in the code. This
  commit replaces all `NaN` values with `null`.

fredmanglis committed:
Return the correlation method used

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Return the correlation method used

fredmanglis committed:
Fix sorting

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Update the sorting algorithm, for literature and tissue correlations so that
  it sorts the results by the correlation value first then by the p-value
  next.

fredmanglis committed:
Fix linting errors

fredmanglis committed:
Fix typing errors

fredmanglis committed:
Encode the data to JSON and set the status code

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Encode bytes objects to string
* Encode NaN values to "null"
* gn3/api/correlation.py:

fredmanglis committed:
Add API endpoint for partial correlations

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Add an API endpoint for the partial correlation.
* gn3/api/correlation.py:

Alexanderlacuna committed:
Refactor wgcna (#63)

* add r as a gn3 input

* calculate powers from user input

* fix merge conflict

BonfaceKilz committed:
traits.py Delete a record only if it exists

In the case when the user tries to delete the same data twice, prior to this
commit, an error was being generated.  This commit remedies this by checking
if a record exists prior to deleting it.

fredmanglis committed:
Rework database functions to fetch multiple items

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* In an attempt to optimise the performance of the partial correlations
  feature, this commit reworks some database access functions to fetch
  multiple items from the database, per query, unlike their original forms
  which would fetch a single item per query.

  This reduces queries to the database, and should hopefully improve the
  responsiveness of the partial correlations feature.

arunisaac committed:
pylintrc: Disable fixme check.

fixme notes such as TODO, FIXME, etc. are good practice, and should not be
discouraged.

* .pylintrc (disable): Add fixme.

arunisaac committed:
db: Fix wrong continued indentation.

* gn3/db/datasets.py (dataset_metadata): Fix wrong continued indentation.

arunisaac committed:
db: Fix sparql_query return type.

* gn3/db/datasets.py: Import List from typing.
(sparql_query): Set return type to List[Dict[str, Any]].

arunisaac committed:
db: Assist mypy with explicit type declaration.

* gn3/db/datasets.py (dataset_metadata): Assist mypy by explicitly declaring
type of return dictionary.

fredmanglis committed:
Convert NaN to None

Issue: https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi
Comment:
https://github.com/genenetwork/genenetwork3/pull/67#issuecomment-1000828159

* Convert NaN values to None to avoid possible bugs with the string replace
  method used before.

fredmanglis committed:
Silence `duplicate-code` linting notifications

fredmanglis committed:
Add optimised entry for partial correlations

Issue:

* Function
  `gn3.computations.partial_correlations_optimised.partial_correlations_entry`
  is a copy of the
  `gn3.computations.partial_correlation.partial_correlations_entry`
  function that is optimised for better performance.

  The optimised function is intended to replace the unoptimised one, but it is
  included in this commit for comparison purposes, and to maintain some
  historical context for doing it this way.

fredmanglis committed:
Replace unoptimised function with optimised one

Issue:
https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

* Replace unoptimised function with one optimised to give better performance.

  The optimisation done here is to fetch multiple items/traits from the
  database per query, rather than the original form, which fetched a single
  item/trait from the database per query.

pjotrp committed:
guix_deploy

pjotrp committed:
guix_deploy

fredmanglis committed:
Fix error in query: Replace '=' with 'IN'

fredmanglis committed:
Remove all pairs with 'None' as the value

* Remove all key-value pairs whose value is None.

fredmanglis committed:
Convert keys to string

* Some traits have a name composed of all numerals, which leads to the names
  being interpreted as numbers. This commit forces them to string to avoid
  subtle bugs where the code fails.

fredmanglis committed:
Indicate that string is an f-string

* The string had the f-string syntax to format the values to be inserted into
  the string, but was missing the 'f' before the opening quotes to signify to
  python that this was an f-string. This commit fixes that.

fredmanglis committed:
Surround statement with parens for readability

fredmanglis committed:
Use the correct letter case for the keys

* Use the correct case for the keys inorder to retrieve the correct values.

BonfaceKilz committed:
traits.py: Return early during an insert if the give record exists

Sometimes, a user will try to insert data twice, on in some instances, 2
different users will attempt the same inserts of the same records separately.
In such cases, ignore the insert, and return early.

BonfaceKilz committed:
traits.py: Don't use "value" to check if a record exists

The problem with using the "value" record is that it's a floating point
number.  See
<https://www.bonfacemunyoki.com/post/2021-10-21-comparing-floating-point-numbers/>
on why comparing floating point numbers can be an issue.

fredmanglis committed:
Fix dataset retrieval query for `Geno` traits

Issue: https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

fredmanglis committed:
Check for path existence

Issue: https://github.com/genenetwork/gn-gemtext-threads/blob/main/topics/gn1-migration-to-gn2/partial-correlations.gmi

BonfaceKilz committed:
test_traits.py: Update failing test

BonfaceKilz committed:
Fix pylint errors

BonfaceKilz committed:
Merge branch 'main' into fix/check-for-duplicates-before-deletions-or-insertions

BonfaceKilz committed:
Merge pull request #69 from genenetwork/fix/check-for-duplicates-before-deletions-or-insertions

Fix/check for duplicates before deletions or insertions

BonfaceKilz committed:
Merge pull request #64 from jgarte/type-hint-normalize-values

Adds type hint for normalize_values function

fredmanglis committed:
Deduplicate query to fetch data for 'Publish' traits

The queries run in the `get_trait_csv_sample_data` and
`retrieve_publish_trait_data` functions in the `gn3.db.traits` module were
mostly similar. This commit changes that, by making the
`get_trait_csv_sample_data` function make use of the results from calling the
`retrieve_publish_trait_data` function.

fredmanglis committed:
Indent code correctly

Indent the code correctly.

fredmanglis committed:
Update return type. Remove duplicate import.

The function is a generator function, since it uses a `yield` statement, and
thus returns a generator object, that contains a tuple object. This fixes
that. We also remove a duplicate import.

Alexanderlacuna committed:
fixes for ctl script:generate ctl ctl_significant json

Alexanderlacuna committed:
generate required json data for ctl api

Alexanderlacuna committed:
read stdout data;handle exceptions

Alexanderlacuna committed:
generate ctl plots for each trait

Alexanderlacuna committed:
process ctl plot data img

Alexanderlacuna committed:
generate network

Alexanderlacuna committed:
function override:fix target specific output file for network

Alexanderlacuna committed:
minor refactoring

jgarte committed:
Use guix shell shortcut with guix.scm file

arunisaac committed:
README: Update link to CI badge.

The CI has been moved to penguin2, and this is accompanied by a URI change.

* README.md: Update link to CI badge.

arunisaac committed:
README: Add pylint and mypy CI badges.

We now have separate CI jobs for the pylint and mypy tests.

* README.md: Add pylint and mypy CI badges.

Alexanderlacuna committed:
fix comprehension list

Alexanderlacuna committed:
add new json output

Alexanderlacuna committed:
new line fix

Alexanderlacuna committed:
use user input significance

Alexanderlacuna committed:
fix for parametric input

Alexanderlacuna committed:
match case string input to boolean

Alexanderlacuna committed:
pep8 formatting

Alexanderlacuna committed:
refactor code for invoking ctl script

Alexanderlacuna committed:
return 401 on request fail

Alexanderlacuna committed:
remove comments

Alexanderlacuna committed:
pep8 formatting

Alexanderlacuna committed:
response object error fix

zsloan committed:
Fix retrieve_publish_trait_data query

The PublishFreeeze table is actually unnecessary for this query, since
the group ID (inbred_set_id) should be passed in and that ID is in the
PublishXRef table (so no neeed to join with PublishFreeze)

zsloan committed:
Fix bug where sample values of 0 were being treated as False

In line 91 of gn3/db/traits.py, there was an if statement "if
record[key] else 'x'" that was treating values of 0 as False, so I
changed it to explicitly check that values aren't None

zsloan committed:
Remove PublishFreeze from retrieve_publish_trait_info query

The PublishFreeze table isn't necessary in phenotype queries, since
PublishFreeze.Id = InbredSet.Id (for the purposes of identifying traits,
at least)

BonfaceKilz committed:
Revert "Use guix shell shortcut with guix.scm file"

This reverts commit f676c291967f8a81836b73c5a5dcd3c65e02552c.

zsloan committed:
Fix R/qtl covar bug

The rqtl_wrapper script was throwing an error when only a single
categorical covariate was used. This is apparently because "covars[,name]"
throws an error in such a situation. Using just "covars" in such a
situation prevents the error. So I just added an if statement checking
the number of covariates. There might be some better way to deal with
this in R, but this is the best I could come up with.

fredmanglis committed:
Remove unnecessary computation

In Python3 when slicing,

     seq[:min(some_val, len(seq))] == seq[:some_val]

because Python3 will just return a copy of the entire sequence if `some_val`
happens to be larger/greater than the length of the sequence.

This commit removes the unnecessary call to `min()`

fredmanglis committed:
Use multiprocessing to speed up computation

This commit refactors the code to make it possible to use multiprocessing to
speed up the computation of the partial correlations.

The major refactor is to move the `__compute_trait_info__` function to the
top-level of the module, and provide to it all the other necessary context via
the new args.

fredmanglis committed:
Give sorting functions more descriptive names

BonfaceKilz committed:
db: Fetch correct sample data

fredmanglis committed:
Remove multiprocessing for stability

Web servers are long-running processes, and python is not very good at
cleaning up after itself especially in forked processes - this leads to memory
errors in the web-server after a while.

This commit removes the use of multiprocessing to avoid such failures.

BonfaceKilz committed:
db: traits: Insert "count" instead of "error" in "NStrain" table

BonfaceKilz committed:
db: traits: Allow insertions of "0" in resp. sample values tables

When editing values from "x" to "0"(or any other value) when editing data, an
"update" statement was being run; thereby no new value was being inserted.  To
the end user, modifying an "x" value to something else meant that no value was
being inserted.  This commit fixes that by doing an insert whenever a change
from "x" to "0" is performed.

* gn3/db/traits.py (update_sample_data): Add insert statements whenever an
"update" statement returns a 0 row-count.

fredmanglis committed:
Merge iterations to remove unnecessary computations

Do all the work in a single iteration to avoid unnecessary iterations that
hamper performance.

BonfaceKilz committed:
db: traits: Enable deletion of individual sample values

If a user replaces an individual value with an "x", delete that date entry
from the respective table.  Deletion here is the only option since by default
the Nstrain, PublishData and PublishSE don't accept null values.  Note that
deleting all 3 values is equivalent to removing the sample from the CSV file.

* gn3/db/traits.py (update_sample_data): If a value is "x", delete it from the
respective table.

fredmanglis committed:
Remove obsolete `PublishFreeze` table

arunisaac committed:
github: Remove GitHub actions.

We have our own laminar CI and no longer need GitHub actions.

* .github/workflows: Delete directory.

BonfaceKilz committed:
db: Extend csv query to fetch case attributes

* gn3/db/traits.py (get_trait_csv_sample_data): Fetch case attribute data if
it exists.

fredmanglis committed:
Add property tests for `dictify_by_samples`

Add property tests using pytest and hypothesis to test that the expected
properties hold for the
`gn3.computations.partial_correlations.dictify_by_samples`
function.

fredmanglis committed:
Quote shell variables to prevent globbing

Quote the shell variables to prevent globbing and word splitting.
Deactivate this check for the specific lines that require intentional word
splitting

fredmanglis committed:
Add scaffolding to run different types of tests

Enable running commands:

    - `python3 setup.py unit_check`: run the unit tests
    - `python3 setup.py integration_check`: run integration tests
    - `python3 setup.py performance_check`: run performance tests

arunisaac committed:
sql: Remove database mapping code.

These tools have been greatly improved and moved to a new home at
https://git.genenetwork.org/arunisaac/dump-genenetwork-database

* sql/map-database.sh, sql/schema-from-in-db-documentation.org,
sql/schema-original.sql, sql/schema.png, sql/schema.sql, sql/schema.svg: New
files.

fredmanglis committed:
Provide custom class to run tests

fredmanglis committed:
Use pytest's "mark" feature to categorise tests

Use pytest's `mark` feature to explicitly categorise the tests and run them
per category

fredmanglis committed:
Create client fixture for integration tests

Create a client fixture to help with the integration tests

fredmanglis committed:
Test partial correlations endpoint with missing data in POST request

Add a test for the partial correlations endpoint, with:
- no data in the request
- missing items in the data

Fix the bugs caught by the test

BonfaceKilz committed:
Revert "db: Fetch correct sample data"

This reverts commit 710769e84b3bc6a2bdd66effdbac0659272ed511.

BonfaceKilz committed:
Append case attributes to csv data if they exist

BonfaceKilz committed:
Fix pylint error

fredmanglis committed:
Test partial correlations endpoint with non-existent primary traits

Test that the partial correlations endpoint responds with an appropriate
"not-found" message and the corresponding 404 status code in the case where a
request is made and the primary trait requested for does not exist in the
database.

Summary of the changes in each file:
* gn3/api/correlation.py: generalise the building of the response
* gn3/computations/partial_correlations.py: return with a "not-found" if the
  primary trait does not exist in the database
* gn3/db/partial_correlations.py: Fix a number of bugs that led to exceptions
  in the case that the primary trait did not exist
* pytest.ini: register a `slow` pytest marker
* tests/integration/test_partial_correlations.py: Add a new test to check for
  an appropriate 404 response in case of a primary trait that does not exist
  in the database.

fredmanglis committed:
Remove code trying to query non-existent `TempFreeze` table

The code was migrated from GN1 with a faulty assumption that all trait types
have a corresponding `*Freeze` table in the database. This assumption is not
true for the `Temp` traits.

This commit removes the buggy code.

fredmanglis committed:
Remove erroneous test based on wrong assumptions

Related to commit 75dcfe295af57b16428c586cc11dbaa827a5feba

This commit removes the related test that was checking for the wrong thing.

fredmanglis committed:
Test partial corrs endpoint with non-existing control traits

Test that if the endpoint is queried and not a single one of the control
traits exists in the database, then the endpoint will respond with a
404 (not-found) status code.

Summary of changes:
* gn3/computations/partial_correlations.py: Check whether any control trait is
  found. If none is found, return "not-found" message.
* gn3/db/partial_correlations.py: Fix bug in Geno query.
* tests/integration/test_partial_correlations.py: Add test for non-existing
  control traits. Rename function to make it clearer what it is testing
  for. Remove obsoleted comments.

fredmanglis committed:
Test partial corrs API with mix of existing and non-existing control traits

Test that the partial correlations endpoint handles a mix of existing and
non-existing control traits gracefully and issues a warning to the user.

Summary of changes:
* gn3/computations/partial_correlations.py: Issue a warning for all
  non-existing control traits
* gn3/db/partial_correlations.py: update queries - use `INNER JOIN` for tables
  instead of comma-separated list of tables
* tests/integration/conftest.py: Add `db_conn` fixture to provide a database
  connection to the tests. This will probably be changed in the future to
  connect to a temporary database for tests.
* tests/integration/test_partial_correlations.py: Add test to check for
  correct behaviour with a mix of existing and non-existing control traits

fredmanglis committed:
Add test to ensure samples are read correctly from .geno files

* tests/unit/db/test_genotypes2.py: New file

fredmanglis committed:
Fix a myriad of linter issues

* Use `with` in place of plain `open`
* Use f-strings in place of `str.format()`
* Remove string interpolation from queries - provide data as query parameters
* other minor fixes

fredmanglis committed:
Fix minor issues introduced while fixing linting errors

BonfaceKilz committed:
db: Fix error in SQL query

* gn3/db/traits.py (get_trait_csv_sample_data): Update SQL to fix runtime
errors.

fredmanglis committed:
Run partial correlations in an external process

Run the partial correlations code in an external python process decoupling it
from the server and making it asynchronous.

Summary of changes:
* gn3/api/correlation.py:
  - Remove response processing code
  - Queue partial corrs processing
  - Create new endpoint to get results
* gn3/commands.py
  - Compose the pcorrs command to be run in an external process
  - Enable running of subprocess commands with list args
* gn3/responses/__init__.py: new module indicator file
* gn3/responses/pcorrs_responses.py: Hold response processing code extracted
  from ~gn3.api.correlations.py~ file
* scripts/partial_correlations.py: CLI script to process the pcorrs
* sheepdog/worker.py:
  - Add the *genenetwork3* path at the beginning of the ~sys.path~ list to
    override any GN3 in the site-packages
  - Add any environment variables to be set for the command to be run

BonfaceKilz committed:
Add methods for working with csv data

gn3/csvcmp.py: New file
(create_dirs_if_not_exists): From a list of dirs, create them if they don't
exist.
(remove_insignificant_edits): Given a dict with a "Modification" key, remove
edits with "delta < ε".
(csv_diff): Generate a csv_diff using the "csvdiff" tool packaged in guix.

tests/unit/test_csvcmp.py: Add some tests for "gn3/csvcmp.py"

BonfaceKilz committed:
Store columns in the output dict

When inserting, deleting, or editing case-attributes, we need the column
headers in order to be able to know identify the attribute of interest.

* gn3/csvcmp.py (csv_diff): Add extra "Column" key in returned dict.

arunisaac committed:
gn3: Explicitly specify UTF-8 to be the file encoding.

When the encoding is not specified explicitly, the system default encoding is
used. This is not recommended.

* gn3/computations/ctl.py (call_ctl_script),
gn3/computations/gemma.py (generate_pheno_txt_file),
gn3/computations/parsers.py (parse_genofile),
gn3/computations/partial_correlations.py (partial_correlations_fast),
gn3/computations/rqtl.py (process_rqtl_output, process_perm_output),
gn3/computations/wgcna.py (dump_wgcna_data, call_wgcna_script),
gn3/fs_helpers.py (jsonfile_to_dict): Explicitly specify UTF-8 to be the file
encoding.
*
tests/unit/computations/test_gemma.py (TestGemma.test_generate_pheno_txt_file),
tests/unit/computations/test_wgcna.py (TestWgcna.test_create_json_file): Test
for call to open with encoding='utf-8' argument.

arunisaac committed:
gn3: computations: Call Popen with context manager.

Context managers should be preferred when allocating resources.

* gn3/computations/wgcna.py (stream_cmd_output): Call Popen with context
manager.

BonfaceKilz committed:
Replace "all" with "and"

* gn3/csvcmp.py (remove_insignificant_edits): "all" evaluates all elements and
throws an error if when `abs(float(x) - float(y)) < epsilon` is processed. Use
"and" instead because of it's short-circuiting behaviour.

BonfaceKilz committed:
Create new method for filling csv with a default value

* gn3/csvcmp.py (fill_csv): Given a csv text with uneven or incomplete fields
whole length are less than width, fill them with a value which defaults to
"x".

* tests/unit/test_csvcmp.py (test_fill_csv): Test cases for the above.

BonfaceKilz committed:
Fill CSV text if there are non-even rows

Should you try to run `csvdiff` against 2 csv files with either file having a
non-even columns, there will be an error. As such, the csv files need to be
"filled" before running `csvdiff`.

* gn3/csvcmp (csv_diff): For non-even rows in the csv files, fill the csv
rows.

BonfaceKilz committed:
Test edges cases for csv files when running csvdiff

* tests/unit/test_csvcmp.py (test_csv_diff): Delete it.
(test_csv_diff_same_columns): Test csv_diff against csv texts with the same
columns.
(test_csv_diff_different_columns): Test csv texts against csv texts with
different varying columns.

BonfaceKilz committed:
Don't add extra key "Column" to dict if there are no changes

gn3/csvcmp.py (csv_diff): If the diff is empty, don't add an extra key
"Column" to the dictionary.
tests/unit/test_csvcmp (test_csv_diff_only_column_change): Add test-case for
the above.

fredmanglis committed:
Do incremental backoff if there are no jobs on the queue

Since the worker polls a queue for jobs, it can get into a busy poll. This was
the reason that there was a delay of 0.1 seconds between each poll instance.

This commit takes this a little further by doing an incremental backoff, where
it waits longer and longer after each poll where it does not find a job
available, up to an arbitrary maximum.

BonfaceKilz committed:
Move operations on sample_data to it's own module

BonfaceKilz committed:
Create a new function for retrieving strain_id and publishdata_id

* gn3/db/sample_data.py: Import Any, Tuple.
(get_sample_data_ids): New function that fetches the strain_id and
publishdata_id of a given data point.
(update_sample_data): Use `get_sample_data_ids`.
(delete_sample_data): Ditto.
(insert_sample_data): Ditto.

zsloan committed:
Fix issue where 0's were treated as False for the primary trait in
correlations

In the original version of the if statement* I believe it was
interpreted as "if a_val and (b_val is not None)". This caused
values of 0 for a_val (the primary trait's values) to be evaluated as
False.

I changed it to compare both a_val and b_val to None. This seems to have
fixed the issue.

* if (a_val and b_val is not None)

BonfaceKilz committed:
Fetch InbredSetId

* gn3/db/sample_data.py (get_sample_data_ids): Extend to also fetch
InbredSetId.
(update_sample_data): Discard the returned value of InbredSetId.
(delete_sample_data): Ditto.

BonfaceKilz committed:
Allow inserting case-attribute data during inserts

* gn3/db/sample_data.py (insert_sample_data): Modify this function to allow
inserting case-attribute values.

BonfaceKilz committed:
Allow deleting case-attribute data during deletion

* gn3/db/sample_data.py (delete_sample_data): Modify this function to allow
deleting case-attribute values.

BonfaceKilz committed:
Extract a strain name given a csv string and it's header

* gn3/csvcmp.py (extract_strain_name): New function.
* gn3/db/sample_data (delete_sample_data): Use the aforementioned function.
(insert_sample_data): Ditto.
* tests/unit/test_csvcmp: Test cases for above.

BonfaceKilz committed:
Fetch id's separately for the insertion edge-case

* gn3/db/sample_data (get_sample_data_ids): Add an extra condition that caters
for inserts: during inserts, joins won't work when fetching the strain_id,
publishdata_id, and strain_name. In this case, just create 2 separate queries
to do that work.

BonfaceKilz committed:
Remove test cases related to sample data

Most of this functions were moved to their own module.

BonfaceKilz committed:
Fill in empty values in csv text with: "x"

* gn3/csvcmp.py (fill_csv): Update this function to allow empty lists to be
filled with the default value(set in the args).
* tests/unit/test_csvcmp.py (test_fill_csv): Update test to capture above.

BonfaceKilz committed:
Check whether publish data already exists before inserting

* gn3/db/sample_data.py (insert_sample_data): If data already exists in the
table, do not attempt an insert; otherwise, an error will be generated.

BonfaceKilz committed:
Remove dead code

BonfaceKilz committed:
Remove duplicate params

* gn3/db/sample_data.py (insert_sample_data)[__insert_case_attribute]: Remove
extra parameters.

BonfaceKilz committed:
Explicitly get CaseAttributeId and fix broken sql query

* gn3/db/sample_data.py (insert_sample_data): Use correct query string. Also,
use CaseAttributeId to determine whether case-attributes were inserted. If so,
do not attempt an insert.

BonfaceKilz committed:
Fix faulty SQL query string when deleting case-attributes

BonfaceKilz committed:
Make `_map` a constant

* gn3/db/sample_data.py: Now constant, `_MAP`.
(delete_sample_data)[__delete_data]: Replace `_map` with `_MAP`.
(insert_sample_data)[__insert_data]: Ditto.

fredmanglis committed:
Add endpoint for checking state of external processes

Long-running computations are handed off to external processes. This avoids
timeouts in the webserver, and also reduces chances of instability of the
webserver.

The results of these long-running computations are needed eventually, so this
commit provides a way to check for the state of the computation, and the
results if any.

BonfaceKilz committed:
Remove check for inserted data when inserting individual data

* gn3/db/sample_data.py (insert_sample_data)[__insert_data]: Move check to the
main body. With this check here, you have 3 redundant checks. For a successful
insert, it will insert the first value to the `PublishData` table and ignore the
rest of the inserts.

fredmanglis committed:
Document basic partial correlations dependencies

* Document the things that the partial correlations feature depends on to work

fredmanglis committed:
Automatically decode Redis strings

BonfaceKilz committed:
Add test cases for inserting and deleting data

* tests/unit/db/test_sample_data.py (test_insert_sample_data): Test inserting
data.
(test_delete_sample_data): Test deleting data.

fredmanglis committed:
Create database connections within context managers

Use the `with` context manager to open database connections, so as to ensure
that those connections are closed once the call is completed. This hopefully
avoids the 'too many connections' error

fredmanglis committed:
Fix tests, and issues caught by tests

Fix some issues caught by tests due to changes introducing the hand-off of the
partial correlations computations to an external process

Fix some issues due to the changes that introduce context managers for
database connections

Update some tests to take the above two changes into consideration

fredmanglis committed:
Remove unused function and its tests

BonfaceKilz committed:
Create action dict that's created when updating data

* gn3/db/sample_data.py (__extract_actions): An update on a vector of data can
contain: inserts, deletes and updates. This functions extracts these actions
during an update.
* tests/unit/db/test_sample_data.py (test_extract_actions): Add test-case for
the above.

BonfaceKilz committed:
Update how data is updated by re-using existing functions

* gn3/db/sample_data.py (get_sample_data_ids): Re-use "delete_sample_data" and
"insert_sample_data" when updating data; and also add logic for updating
modified data.
* tests/unit/db/test_sample_data.py: Add tests for the above.

BonfaceKilz committed:
Add missing return type-annotations

* tests/unit/db/test_sample_data.py (delete_sample_data): Add missing return
type for type annotations.
(insert_sample_data): Ditto.

BonfaceKilz committed:
Apply auto-pep8 to sample_data.py and it's test file

BonfaceKilz committed:
Append the strain name when extracting "actions"

* gn3/db/sample_data.py (__extract_actions): During updates, make sure that
the strain name is part of the returned string when extracting "actions".
* tests/unit/db/test_sample_data.py: Add test cases for the above.

BonfaceKilz committed:
Store the first element as strain_id

fredmanglis committed:
Fix some linting issues

fredmanglis committed:
Fix typing errors

BonfaceKilz committed:
Delete noisy "print" statement

BonfaceKilz committed:
Compose csv-diff command within single quotes

* gn3/csvcmp.py (csv_diff): Use single quotes. There was a change in 6d39c92
that broke this.

BonfaceKilz committed:
README: Replace "unit-test" instructions with "pytest"

BonfaceKilz committed:
Fix pylint issues

BonfaceKilz committed:
Fix mypy issues

BonfaceKilz committed:
Remove unused imports

BonfaceKilz committed:
Fix pylint errors in unit_tests

BonfaceKilz committed:
Get all permissible column data

* gn3/csvcmp.py: Import "Any" and "List".
(get_allowable_sampledata_headers): New function.
* tests/unit/test_csvcmp: Import "get_allowable_sampledata_headers".
(test_get_allowable_csv_headers): Test case for the above.

BonfaceKilz committed:
Given a csv text and permissible headers, extract invalid headers

* gn3/csvcmp.py (extract_invalid_csv_headers): New function.
* tests/unit/test_csvcmp.py: Import "extract_invalid_csv_headers".
(test_extract_invalid_csv_headers_with_some_wrong_headers): Test case for the
above.

BonfaceKilz committed:
Only loop through the diff's modifications if it exists

BonfaceKilz committed:
Sort import lines

BonfaceKilz committed:
Dummy White Space commit to fix laminar

Alexanderlacuna committed:
Feature/refactored pca (#79)

* compute zscore function

* test case for computing zscore

* function to compute pca

* generate scree plot data

* generate  new pca trait data from zscores and eigen_vec

* remove redundant functions

* generate factor loading table data

* generate pca temp dataset dict

* variable naming and error fixes

* unit test for processing factor loadings

* minor fixes for generating temp pca dataset

* pass datetime as argument to generate_pca temp dataset function

* add unittest  for caching pca datasets

* cache temp datasets

* ignore missing imports for sklearn

* mypy fixes

* pylint fixes

* refactor tests for pca

* remove ununsed imports

* fix for generating pca traits vals

* mypy and code refactoring

* pep8 formatting and add docstrings

* remove comments /pep8 formatting

* sort eigen vectors based on  eigen values

* minor fix for zscores

* fix for rounding variance ratios

* refactor tests

* rename module to pca

* rename datasets to traits

* fix failing tests

* fix caching function

* fixes return x and y coordinates for scree plot

* expand exception scope

* fix for deprecated numpy.matrix function

* fix for failing tests

* pep8 fixes

* remove comments

* fix merge conflict

* pylint fixes

* rename module name to test_pca

BonfaceKilz committed:
Create new function for cleaning individual fields in csv text

* gn3/csvcmp.py (clean_csv_text): New function.
* tests/unit/test_csvcmp.py: Import "csv_text".
(test_clean_csv_text): Test case for the above.

BonfaceKilz committed:
Clean all csv fields before diffing

There was a subtle bug where "csvdiff" generated an error related to
"different column headings" caused something akin to diffing: "a, b \n, ..."
with "a, b\n, ...".

* gn3/csvcmp.py (csv_diff): Clean csv texts before any diffing.
* tests/unit/test_csvcmp.py (test_csv_diff_same_columns): Modify test case to
capture aforementioned bug.

arunisaac committed:
README: Update CI badge subdomain.

The CI has been migrated from penguin2.genenetwork.org to ci.genenetwork.org.

* README.md: Update CI badge subdomain.

zsloan committed:
Change order of if statements for running genoprob command

Now it checks for pairscan first, just in case interval ends up being
passed (which is an irrelevant parameter for pairscan)

Also added a couple more verbose prints

zsloan committed:
Merge branch 'feature/add_rqtl_pairscan' of https://github.com/zsloan/genenetwork3 into feature/add_rqtl_pairscan
